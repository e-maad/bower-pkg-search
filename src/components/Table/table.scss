.table {
  width: 100%;
  border-collapse: collapse;
  color: var(--color-brown);
  & thead th {
    border-bottom: 1px solid #eee;
    padding: 5px 0 15px 0;
    text-align: left;
    min-width: 80px;
    text-decoration: none;
    color: #008ec4;
    text-align: left;
    cursor: pointer;
  }
  & tbody tr {
    border-bottom: 1px solid #eee;
    &:last-child {
      border-bottom: none;
    }
  }
  & td span {
    word-break: break-all;
    white-space: normal;
  }
}
