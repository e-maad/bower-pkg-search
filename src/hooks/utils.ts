import { config } from "../config";
import { FetchModuleParams, Module } from "./types";

export const getSearchURL = ({
  search: q,
  sortBy,
  itemsPerPage,
  page,
}: FetchModuleParams) => {
  const { apiKey: api_key, apiURL } = config;

  const searchParams = new URLSearchParams({
    q,
    per_page: itemsPerPage.toString(),
    page: page.toString(),
    api_key,
    ...(sortBy && { sort: sortBy.toString() }),
  });

  return `${apiURL}search?${searchParams}`;
};

export const mapSearchResult = (result: any[]): Module[] => {
  return result.map((item) => ({
    name: item.name,
    url: item.repository_url,
    description: item.description,
    owner: item.repository_url?.split("/").splice(-2, 1) || "",
    stars: item.stars,
    homepage: checkHomepageURL(item.homepage),
  }));
};

export const checkHomepageURL = (url: string) => {
  if (!url) return "";

  if (!/^(f|ht)tps?:\/\//i.test(url)) {
    url = "http://" + url;
  }
  return url;
};
