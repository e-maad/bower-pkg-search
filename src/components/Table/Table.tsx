import React from "react";
import "./table.scss";
import Skeleton from "../Skeleton/Skeleton";

export interface Column<DataType> {
  name: string;
  key: keyof DataType;
  sortable?: boolean;
  render?: (row: DataType) => JSX.Element;
  width?: React.CSSProperties["width"];
}

interface TableProps<DataType> {
  columns: Column<DataType>[];
  data: DataType[];
  sortBy?: keyof DataType;
  loading?: boolean;
  onSort: (sortBy?: keyof DataType) => void;
}

export default function Table<DataType>({
  data,
  columns,
  loading,
  sortBy,
  onSort,
}: TableProps<DataType>) {
  const handleSort = ({
    sortable,
    key,
  }: Pick<Column<DataType>, "key" | "sortable">) => {
    if (!sortable) return;

    if (sortBy === key) {
      onSort();
    } else {
      onSort(key);
    }
  };

  return (
    <table className="table">
      <thead>
        <tr>
          {columns.map(({ name, width, key, sortable }) => (
            <th
              style={{ width }}
              key={String(key)}
              onClick={() =>
                handleSort({
                  sortable,
                  key,
                })
              }
            >
              {name}
              {sortable && sortBy === key && <span>&#8595;</span>}
            </th>
          ))}
        </tr>
      </thead>
      {loading ? (
        <tbody>
          <tr>
            {columns.map(({ key }) => (
              <td key={String(key)}>
                <Skeleton />
              </td>
            ))}
          </tr>
        </tbody>
      ) : (
        <tbody data-testid="table-body">
          {data.map((row, index) => (
            <tr key={index}>
              {columns.map(({ key, render }) => (
                <td key={index + String(key)}>
                  {render ? render(row) : <span>{String(row[key])}</span>}
                </td>
              ))}
            </tr>
          ))}
        </tbody>
      )}
    </table>
  );
}
