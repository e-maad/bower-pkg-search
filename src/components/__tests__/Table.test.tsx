import { render } from "@testing-library/react";
import Table from "../Table/Table";
import { TestData, getTestData, testDataColumns } from "./utils";

describe("<Table />", () => {
  it("should validate the snapshot", async () => {
    const component = render(
      <Table<TestData>
        data={[getTestData()]}
        columns={testDataColumns}
        loading={false}
        onSort={jest.fn()}
      />
    );
    expect(component).toMatchSnapshot();
  });
});
