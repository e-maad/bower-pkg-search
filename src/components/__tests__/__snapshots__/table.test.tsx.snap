// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should validate the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="table"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Type
            </th>
          </tr>
        </thead>
        <tbody
          data-testid="table-body"
        >
          <tr>
            <td>
              <span>
                Name
              </span>
            </td>
            <td>
              <span>
                Type
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="table"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Type
          </th>
        </tr>
      </thead>
      <tbody
        data-testid="table-body"
      >
        <tr>
          <td>
            <span>
              Name
            </span>
          </td>
          <td>
            <span>
              Type
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
